support both sqlalchemy engines and connections

graingert opened the pull request:
Fixes #7877

artemyk commented:
This looks good and would also resolve #7877 I believe.
I think some tests will probably be requested.

jorisvandenbossche commented:
@graingert Thanks for working on this!

Certainly look also to the discussion in #7877. At the end there is link to a patch of @tr11 which does something similar.

Some quick thoughts:

- indeed, as @artemyk mentions, this needs some tests. 
- I think this will not work for `to_sql` and `read_sql_table`? (how you implemented it now, but is should work for those as well)

jorisvandenbossche commented:
@graingert Some of the tests keep on running. I already restarted them a few times, so it seems this is caused by this PR. Did you try to run the tests locally? (only the sql tests is enough)

Commit:
support both sqlalchemy engines and connections Fixes #7877

update pymysql to 0.6.3 to avoid cursor bugs

Add documentation and tests for SQLAlchemy connectables

explicit reference to connection/engine in docs

Temporary table test

pass compile the connectable

graingert commented on pandas/io/tests/test_sql.py:
```diff
@@ -2180,6 +2237,7 @@ def setUp(self):
         except:
             pass
         else:
+            self.addCleanup(self._try_close_db)
```
For the tests to pass on my machine I need to close the database connection after each test.

graingert commented:
@jorisvandenbossche I've debugged all the tests.

Once the travis builds pass I'll remove `Fix TestXMySQL` and `increase verbosity so I can see what's going wrong`

graingert commented:
I think `Fix TestXMySQL` should probably go in another PR

graingert commented:
@jorisvandenbossche @artemyk I can't replicate the issues in  4b2f087...1037144 do you have any ideas?

graingert commented:
@artemyk @jorisvandenbossche I'm pretty sure I've got it now. 

I'll clean up this PR with a repo tomorrow

graingert commented:
When running with two queries at the same time, if one raises a warning and the other doesn't PyMySQL raises a TypeError.

```python
        proc = """DROP PROCEDURE IF EXISTS get_testdb;
         CREATE PROCEDURE get_testdb ()
 
         BEGIN
             SELECT * FROM test_procedure;
         END"""
```

The DROP PROCEDURE was raising a warning on Travis because the get_testdb procedure was missing, however on my local machine I was not destroying/rebuilding the database each test run which meant I still had the "get_testdb" procedure.

artemyk commented:
@graingert Glad you've figured it out. I'm not sure `execute` necessarily supports multiple statements, for example some DBAPIs (including PyMySQL if I'm not mistaken) have an `executemany` method for this purpose.

graingert commented:
@artemyk executemany is for executing one statement with multiple parameters, not multiple statements.

jorisvandenbossche commented:
@graingert Thanks for the updates! some questions:

* What was in the "Fix TestXMySQL" commit? 
* The problem with the stored procedure, this is only with a connection, not with an engine? In that case, maybe good to keep the original test for engine only (skip when connection), in addition to your adapated test.
* What was the problem with the pymysql version exactly?

graingert commented:
XMySQL is only a problem on my local machine, I've got another PR for that
fix

Using pymysql 0.6.6 this error occurs in whichever test case runs first so
it's fine to run like this on the engine and the connection.

Using pymysql 0.6.3 'has various bug fixes' so I'm not sure what the actual
bug was.
On 29 May 2015 11:06 am, "Joris Van den Bossche" <notifications@github.com>
wrote:

> @graingert <https://github.com/graingert> Thanks for the updates! some
> questions:
>
>    - What was in the "Fix TestXMySQL" commit?
>    - The problem with the stored procedure, this is only with a
>    connection, not with an engine? In that case, maybe good to keep the
>    original test for engine only (skip when connection), in addition to your
>    adapated test.
>    - What was the problem with the pymysql version exactly?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/pandas/pull/10105#issuecomment-106763509>.
>

jorisvandenbossche commented:
There was a problem before with pymysql 0.6.6, see https://github.com/pydata/pandas/issues/9608 (but could never reproduce this locally)

graingert commented:
Not sure what that issue is. This will work on 0.6.3 as well
On 29 May 2015 11:15 am, "Joris Van den Bossche" <notifications@github.com>
wrote:

> There was a problem before with pymysql 0.6.6, see #9608
> <https://github.com/pydata/pandas/issues/9608> (but could never reproduce
> this locally)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/pandas/pull/10105#issuecomment-106765664>.
>

jorisvandenbossche commented:
So if I understand correctly, the problem with the `test_read_procedure` test is caused by pymysql (it works for me with pymysql 0.6.2, but errors with 0.6.6). 
Is there a reason to change the pymysql version on travis? (your second commit) Or can we keep the older version there untill this is fixed on the pymysql side, and then keep the test as it was before.

graingert commented:
There's two bugs. One with pymysql 0.6.1 (dodgy cursors) and one with pymysql 0.6.6 (bad multi statement warnings).

jorisvandenbossche commented:
Ah, yes, but you said that with 0.6.3 the first one (dodgy cursors) is also solved? Then maybe use 0.6.3 instead of 0.6.6, and then the procedure test does not need to be changed?

graingert commented:
@jorisvandenbossche yeah looks like it works fine with 0.6.3

graingert commented:
@jorisvandenbossche yeah the Travis tests pass on 0.6.3 what's your thoughts on a merge?

graingert commented:
@jorisvandenbossche *casual nudge*

jorisvandenbossche commented:
@graingert Thanks for the nudge :-)

This further needs:
- a notice in the whatsnew file (see doc/source/whatsnew)
- update to the docs: both all docstrings and the tutorial docs (see doc/source/io.rst)
- I think we should put in an explicit test that the desired behaviour works (and keeps working). Now it is only tested that the existing tests also run with passing a Connection, but we should add a test (only for when using Connection) that the example you give in the issue your reported works)

jreback commented:
pls rebase on master, some issues have been fixed with the builds

graingert commented:
@jorisvandenbossche done @jreback yep, rebase fixed those issues thanks.

graingert commented:
@jorisvandenbossche @jreback *casual nudge*

graingert commented:
@jorisvandenbossche @jreback hey guys, I really need this feature in production... can you guys merge this if you think it's useful

jreback commented:
@graingert lgtm, but @jorisvandenbossche may have some more comments.

if you are actually running off of master (in production)?, then you can just merge this into your local master. This would be in 0.17.0 which won't be out for prob 2 months.

jorisvandenbossche commented:
Sorry for the slow response. But I will look at it this evening, I promise!

Short comment: ideally, I would like to see a test that now fails (so a test that is only run for the test class that uses a connection, and not an engine), to be sure that with a later refactor this ability is not lost again.

graingert commented:
@jorisvandenbossche that test is added here: https://github.com/pydata/pandas/pull/10105/files#diff-052a926a9b2c33a84e27d51c8fad3466R1372

jorisvandenbossche commented:
Yes, I saw that, but if I understand correctly, this test is ran for both the engine tests as the connection tests? 
I was thinking that this enhancement adds something that was not possible now with engines (a use case for which you need the connection, and can't use the engine)? Eg using a temporary table. So it would be useful to have a test with such a use case, so we ensure that such functionality keeps working.

jorisvandenbossche commented on pandas/io/sql.py:
```diff
@@ -139,7 +139,7 @@ def execute(sql, con, cur=None, params=None):
     sql : string
         Query to be executed
-    con : SQLAlchemy engine or sqlite3 DBAPI2 connection
+    con : SQLAlchemy connectable or sqlite3 DBAPI2 connection
```
I worry a bit that the term 'connectable' is not really known to most users? Maybe just use `engine/connection`?

jorisvandenbossche commented on pandas/io/tests/test_sql.py:
```diff
@@ -1346,9 +1369,35 @@ def test_double_precision(self):
+        """
+        This tests the example raised in issue
+        https://github.com/pydata/pandas/issues/10104
+        """
```
Can you add these comments as normal comments? (using `#`) 
Problem is that otherwise the docstring masks the test name in the output of the testsuite

graingert commented:
Ah I see what you mean

graingert commented:
Ah I see what you mean

jorisvandenbossche commented on pandas/io/tests/test_sql.py:
```diff
@@ -854,6 +857,31 @@ def test_sqlalchemy_type_mapping(self):
+        self.__tx.rollback()
+        self.conn.close()
+        self.conn = self.__engine
+        self.pandasSQL = sql.SQLDatabase(self.__engine)
```
Does this line have to be here?

graingert replied:
not currently, it's just so this class cleans up everything from the superclass

jorisvandenbossche replied:
OK, but before each test is run, the setup will overwrite this anyway again?

graingert replied:
this is so that when the tearDown method gets called in the superclasses the instance is exactly how that superclass set it up in setUp

jorisvandenbossche replied:
ah, yes, OK

jorisvandenbossche commented on pandas/io/sql.py:
```diff
@@ -637,7 +637,8 @@ def exists(self):
         from sqlalchemy.schema import CreateTable
-        return str(CreateTable(self.table).compile(self.pd_sql.engine))
+        engine = self.pd_sql.connectable.engine
+        return str(CreateTable(self.table).compile(engine))
```
`compile` can take an engine or connection, so passing the connectable itself here is OK I think

graingert replied:
good point

jorisvandenbossche commented:
Thanks for adding the temp table test! Gave it a last round, see last comments.

Last question: can you squash the commits? Then I will merge when it is green again

graingert commented:
fixed, rebased, squashed It's party time: :cake: :sparkles:

jorisvandenbossche commented:
@graingert Thanks a lot! (for the contribution, and also for the patience ..)

graingert commented:
Woot, thanks very much!
